native/src/main/java/io/ballerina/lib/data/yaml/utils/JsonTraverse.java

```diff
@@ -0,0 +1,323 @@
+ *
+ * WSO2 LLC. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.lib.data.yaml.utils;
+
+import io.ballerina.lib.data.yaml.common.Types;
+import io.ballerina.lib.data.yaml.parser.ParserUtils;
+import io.ballerina.lib.data.yaml.parser.Values;
+import io.ballerina.runtime.api.PredefinedTypes;
+import io.ballerina.runtime.api.TypeTags;
+import io.ballerina.runtime.api.creators.ValueCreator;
+import io.ballerina.runtime.api.flags.SymbolFlags;
+import io.ballerina.runtime.api.types.ArrayType;
+import io.ballerina.runtime.api.types.Field;
+import io.ballerina.runtime.api.types.IntersectionType;
+import io.ballerina.runtime.api.types.MapType;
+import io.ballerina.runtime.api.types.RecordType;
+import io.ballerina.runtime.api.types.TupleType;
+import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.types.UnionType;
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.utils.TypeUtils;
+import io.ballerina.runtime.api.utils.ValueUtils;
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BError;
+import io.ballerina.runtime.api.values.BMap;
+import io.ballerina.runtime.api.values.BString;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Stack;
+
+/**
+ * Traverse json tree.
+ *
+ * @since 0.1.0
+ */
+public class JsonTraverse {
+
+    private static final ThreadLocal<JsonTree> tlJsonTree = ThreadLocal.withInitial(JsonTree::new);
+
+    public static Object traverse(Object json, BMap<BString, Object> options, Type type, Types.YAMLSchema schema) {
+        JsonTree jsonTree = tlJsonTree.get();
+        try {
+            Object allowDataProjection = options.get(Constants.ALLOW_DATA_PROJECTION);
+            jsonTree.schema = schema;
+            if (allowDataProjection instanceof Boolean) {
+                jsonTree.allowDataProjection = false;
+            } else if (allowDataProjection instanceof BMap<?, ?>) {
+                jsonTree.allowDataProjection = true;
+                jsonTree.absentAsNilableType =
+                        (Boolean) ((BMap<?, ?>) allowDataProjection).get(Constants.ABSENT_AS_NILABLE_TYPE);
+                jsonTree.nilAsOptionalField =
+                        (Boolean) ((BMap<?, ?>) allowDataProjection).get(Constants.NIL_AS_OPTIONAL_FIELD);
```

Comment:
I would prefer to have another class or a record for this user-options and use that.
But this method also ok to me